### PR TITLE
Make aliases faster

### PR DESCRIPTION
**Pull request by bluelhf**
### Description

- Switched `AliasesProvider.materials` set implementation from Java SE `HashSet` to FastUtil `ObjectOpenHashSet`.
- Re-wrote `AliasesParser#fixName(String)`.
    > **Note**  
    > The re-written method is stricter about which kinds of whitespace it strips. This doesn't affect any of our aliases, as the spaces are never used.

---
**Target Minecraft Versions:** any 
**Requirements:** n/a
**Related Issues:** n/a

### CODE REVIEW

**Review by APickledWalrus — COMMENTED** [review]
Should be good in theory, probably just needs some extra testing :)

File: src/main/java/ch/njol/skript/aliases/AliasesParser.java [hunk 1]
```diff
@@ -702,18 +702,56 @@ protected void loadSingleAlias(Map<String, Variation> variations, String item) {
...
+		 *        - The broken pipe is used to separate the singular and plural forms of the alias
+		 *        - We want to allow a space before it for readability
+		 *        - We also don't want to literally include the space in the alias
+		 *
+		 * We also keep track of the first and last non-whitespace characters to trim the string manually.
+		 * Since we're also skipping whitespace, we need to keep track of how many characters we've skipped
+		 * so that our indices for those characters aren't misaligned.
+		 * */
+
+		final StringBuilder sb = new StringBuilder(name.length());
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/aliases/AliasesParser.java:725]: typically we don't include final keywords :)

File: src/main/java/ch/njol/skript/aliases/AliasesParser.java [hunk 2]
```diff
@@ -702,18 +702,56 @@ protected void loadSingleAlias(Map<String, Variation> variations, String item) {
...
+		int lastWhitespace = -1;
+		int stripped = 0;
+
+		for (int i = 0; i < name.length(); ++i) {
+			final char c = name.charAt(i);
+			if (c > ' ') {
+				// The index will be off by the number of characters we've stripped so far
+				final int adjustedIndex = i - stripped;
+
+				if (firstNonWhitespace == -1) firstNonWhitespace = adjustedIndex;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/aliases/AliasesParser.java:738]:
```suggestion
				if (firstNonWhitespace == -1)
					firstNonWhitespace = adjustedIndex;
```
single if statements should be formatted like this :D

File: src/main/java/ch/njol/skript/aliases/AliasesParser.java [hunk 3]
```diff
@@ -702,18 +702,56 @@ protected void loadSingleAlias(Map<String, Variation> variations, String item) {
...
+				lastNonWhitespace = adjustedIndex;
+			} else {
+				final int oldLastWhitespace = lastWhitespace;
+				lastWhitespace = i;
+
+				if ((oldLastWhitespace != -1 && oldLastWhitespace == i - 1)
+					|| (i < name.length() - 1 && name.charAt(i + 1) == '¦')) {
+					++stripped;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/aliases/AliasesParser.java:746]:
```suggestion
				if (
					(oldLastWhitespace != -1 && oldLastWhitespace == i - 1)
					|| (i < name.length() - 1 && name.charAt(i + 1) == '¦')
				) {
					stripped++;
```
I think this formatting is maybe clearer? Also, I don't believe the post/pre increment matters here (I think it ends up as the same thing in this case), so we may prefer to just use a post.

File: src/main/java/ch/njol/skript/aliases/AliasesProvider.java [hunk 4]
```diff
@@ -171,7 +173,7 @@ public AliasesProvider(int expectedCount, @Nullable AliasesProvider parent) {
 		this.aliases = new HashMap<>(expectedCount);
 		this.variations = new HashMap<>(expectedCount / 20);
 		this.aliasesMap = new AliasesMap();
-		this.materials = new ArrayList<>();
+		this.materials = new ObjectOpenHashSet<>();
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/aliases/AliasesProvider.java:176]: What is the default size here? I wonder if we would benefit from using a higher initial capacity value? :)

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
great stuff!!

File: src/main/java/ch/njol/skript/aliases/AliasesProvider.java (hunk 4 above)
Comment (bluelhf) [reply src/main/java/ch/njol/skript/aliases/AliasesProvider.java:176]: The default size is 16 — I left it out as I wasn't sure how many materials we'd end up with. In any case, I did not notice any meaningful performance impact from having to resize the table.

File: src/main/java/ch/njol/skript/aliases/AliasesParser.java (hunk 1 above)
Comment (bluelhf) [reply src/main/java/ch/njol/skript/aliases/AliasesParser.java:725]: It's grown into a habit, sorry.

Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/aliases/AliasesParser.java]: no worries!! probably not a terrible one tbh 🤣

File: src/main/java/ch/njol/skript/aliases/AliasesParser.java [hunk 5]
```diff
@@ -702,18 +702,59 @@ protected void loadSingleAlias(Map<String, Variation> variations, String item) {
...
+
+		StringBuilder sb = new StringBuilder(name.length());
+
+		int firstNonWhitespace = -1;
+		int lastNonWhitespace = -1;
+		int lastWhitespace = -1;
+		int stripped = 0;
+
+		for (int i = 0; i < name.length(); ++i) {
+			final char c = name.charAt(i);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/aliases/AliasesParser.java:733]: you missed one ;)

**Review by APickledWalrus — APPROVED** [review]
Looks good in terms of code, just needs some testing (I may try to do some or find people to run it) :)

**Comment by bluelhf** [comment]
According to testing, the short-circuit breaks comparison in some cases, and fixing it makes it slower than not including it in the first place. I have removed it for the time being.

**Comment by bluelhf** [comment]
Excuse the mess, the kids are learning Git

**Comment by bluelhf** [comment]
A naïve benchmark, times as reported in console by Skript.
Table of benchmarks showing 10 entries without and 10 with optimisations, optimisations taking on average 1.74 times less time.

**Review by TheLimeGlass — APPROVED** [review]

**Pull request merged**